In [4]:
# importing the basic libraries
import numpy as np
import pandas as pd

In [2]:
# lets import a dataset to perform the operations
data = pd.read_csv('Datasets/employee.csv')

# lets check the shape of the dataset
data.shape

(1470, 35)

In [3]:
# lets check the structure of the dataset
data.head(3)

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  

[3 rows x 35 columns]

### The Functions which will help us to identify the null values present in the dataframes

#### The Functions which help us to know whether the data is null or not
* ```isnull()```: Indicates the presence of Missing values in a dataset, return boolean values. i.e., It will return True if the data is missing and False if the data is not missing.
* ```notnull()```: This Function is completely the Opposite of ```isnull()```, It also returns boolean values.

#### The Functions which helps us to compute operations to check  whether all or some data is missing
* ```any()```: It will be used where we want to include all the data for operations and Identification.
* ```all()```: It will be used where we are only checking on some or any of the data for operations or Identification.

In [6]:
# lets first check if there is any null value present in the data
data.isnull().sum().sum()

0

In [7]:
pd.set_option('max_rows', 1470)
data.isnull().sum(axis = 1)

0       0
1       0
2       0
3       0
4       0
5       0
6       0
7       0
8       0
9       0
10      0
11      0
12      0
13      0
14      0
15      0
16      0
17      0
18      0
19      0
20      0
21      0
22      0
23      0
24      0
25      0
26      0
27      0
28      0
29      0
30      0
31      0
32      0
33      0
34      0
35      0
36      0
37      0
38      0
39      0
40      0
41      0
42      0
43      0
44      0
45      0
46      0
47      0
48      0
49      0
50      0
51      0
52      0
53      0
54      0
55      0
56      0
57      0
58      0
59      0
60      0
61      0
62      0
63      0
64      0
65      0
66      0
67      0
68      0
69      0
70      0
71      0
72      0
73      0
74      0
75      0
76      0
77      0
78      0
79      0
80      0
81      0
82      0
83      0
84      0
85      0
86      0
87      0
88      0
89      0
90      0
91      0
92      0
93      0
94      0
95      0
96      0
97      0
98      0
99      0


**We can see that, there are no missing values in the dataset**

* Using ```isnull()``` function returns true, If there any data present in the cell and false if there is no data present in the cell.
&nbsp;

* Using ```isnull().sum()``` function would return the sum of null values present in each of the columns.
&nbsp;

* Using ```isnull().sum().sum()``` function would return the totak number of null values or missing values present in a dataset or dataframe.
&nbsp;

* In the Above Result we can see that every columns present in the dataset are having 0 null values, that means there is no null data in the dataframe.


In [2]:
# lets try to take another dataframe, and try to handle missing values in the dataframe

data = pd.read_csv('melbourne.csv')
data.shape

(23547, 21)

In [3]:
data.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

         Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  03-09-2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  03-12-2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  04-02-2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  04-02-2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  04-03-2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.8014    144.9958  Northern Metropolitan   
1        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
2     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
3        NaN        Yarra  -37.8114    145.0116  Northern Metropolitan   
4     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [19]:
# lets check the basic information about the data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23547 entries, 0 to 23546
Data columns (total 21 columns):
Suburb           23547 non-null object
Address          23547 non-null object
Rooms            23547 non-null int64
Type             23547 non-null object
Price            18396 non-null float64
Method           23547 non-null object
SellerG          23547 non-null object
Date             23547 non-null object
Distance         23546 non-null float64
Postcode         23546 non-null float64
Bedroom2         19066 non-null float64
Bathroom         19063 non-null float64
Car              18921 non-null float64
Landsize         17410 non-null float64
BuildingArea     10018 non-null float64
YearBuilt        11540 non-null float64
CouncilArea      15656 non-null object
Lattitude        19243 non-null float64
Longtitude       19243 non-null float64
Regionname       23546 non-null object
Propertycount    23546 non-null float64
dtypes: float64(12), int64(1), object(8)
memory usage: 3.8+ M

In [5]:
data.shape

(23547, 21)

In [6]:
# lets check whether this dataset is having any null values or not

# first lets check how many rows are having all the null values
data.isnull().all(axis = 1).sum()

0

* The returned value is the total number of missing values in the dataframe
* 66, 918 number of data points are missing that is huge.
* let's also check the details of the missing values.
* In the detailed report we will be able to check how many missing values are associated with each of teh columns.

In [7]:
# lets check the number of missing values in the dataset
# we are check the missing values in the context of columns
data.isnull().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Price             5151
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          4481
Bathroom          4484
Car               4626
Landsize          6137
BuildingArea     13529
YearBuilt        12007
CouncilArea       7891
Lattitude         4304
Longtitude        4304
Regionname           1
Propertycount        1
dtype: int64

* We can see that there are many columns which are having null values, Infact there are some columns where we have huge number of missing values.
* lets also check the percentage of missing values present in the data.



### How to Handle Missing Values from a Dataframe

* **We have already discussed the ways to handle the missing values from the Dataset, but when talking about it in broad sense then only two major choices:**
    * To ```drop/delete``` the missing values
    * To ```impute``` the missinng values
    
* **We have to take decision regarding which of the two options to go with depending on the missing values**.
    * We can drop/delete the missinng values only when the missing percentage is more than 50% or it is not so important.
    * in that case, we have to impute the values using any of the methods listed in the introduction part.
    
lets take some examples using the available dataset to check how to impute or delelete the missing values
    

### Methods we will use to Impute Missing Values

* There are many ways to Impute Missing Values, some of them which are popularly used:
    * ```dropna()```: This Function is used to drops the missing values from a dataset and returns the rest of the dataframe.
    * ```fillna()```: This Function is used to replace the missing value with a specified value.

In [35]:
# lets check the percentage of missing values in the columns of the dataset

percentage_of_missing_values = round(100*(data.isnull().sum()/len(data.index)), 2)
print(percentage_of_missing_values)

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.88
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2         19.03
Bathroom         19.04
Car              19.65
Landsize         26.06
BuildingArea     57.46
YearBuilt        50.99
CouncilArea      33.51
Lattitude        18.28
Longtitude       18.28
Regionname        0.00
Propertycount     0.00
dtype: float64


<p style="color: green;">It can be noticed very easily from the above results that there are many columns have huge percentage of data missing from the dataset. There are columns which are having ```18% to 26%``` of values missing from the data, and also there are some columns which are having ```33% to 56%``` of values missing from the data.

* What action should we take, should we remove them all, or impute.
    * We ```cannot remove the columns having 18% to 26% missing values```, so In this case we will have to impute the missing values using one of the methods.
        * It is very important to impute the missing values for these columns as If we remove these columns from the data having only 18 to 26% of the missing values then we will face huge data loss, In that case, We might lose some of the very important and relevant data for the analysis and results.
    
    * But, ```We can remove the columns having 33 to 56% of missing values```, so In this case will have to drop these columns from the data.
        * It is very important to remove these columns because if we impute these missing values then we can introduce a huge bias into the data, which will be bad for the analysis of the data, It might also return inappropriate and misleading results.

### Deleting the Columns using drop() Function

In [8]:
# removing the three columns having 33% to 56% of missing values in the dataset
data = data.drop('BuildingArea', axis=1)
data = data.drop('YearBuilt', axis=1)
data = data.drop('CouncilArea', axis=1)

# data = data.drop(['Building Area', 'YearBuilt', 'CouncilArea'], axis = 1)

# lets check the columns left after the deletion of above listed columns
data.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'Lattitude', 'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

In [9]:
# count the number of rows having > 5 missing values
# use len(df.index)
len(data[data.isnull().sum(axis=1) > 5].index)

4278

In [10]:
# retaining the rows having <= 5 NaNs
data = data[data.isnull().sum(axis=1) <= 5]

In [11]:
# look at the summary again
round(100*(data.isnull().sum()/len(data.index)), 2)

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.71
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2          1.05
Bathroom          1.07
Car               1.81
Landsize          9.65
Lattitude         0.13
Longtitude        0.13
Regionname        0.00
Propertycount     0.00
dtype: float64

* It can be seen that ```Price Column``` still has large number of missing values i.e., 21% of the data is missing. If we impute these values it will introduce heavy bias into the data and the results will be misleading so It is better to discard the rows where Price is having missing values.

In [12]:
# removing all the rows where there is a missing value in the Price Column
data = data[~np.isnan(data['Price'])]

In [51]:
# look at the summary again
round(100*(data.isnull().sum()/len(data.index)), 2)

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         1.05
Bathroom         1.07
Car              1.76
Landsize         9.83
Lattitude        0.15
Longtitude       0.15
Regionname       0.00
Propertycount    0.00
dtype: float64

* Now, we have Landsize, which is having high percentage of missing values, lets check the range of values present in the Landsize attribute of the data, so that we can impute the missing values present in the Landsize Attribute.

In [52]:
# lets check the description of values in the Landsize Column
data['Landsize'].describe()

count     13603.000000
mean        558.116371
std        3987.326586
min           0.000000
25%         176.500000
50%         440.000000
75%         651.000000
max      433014.000000
Name: Landsize, dtype: float64

* From the above description of the range of values present in the landsize columns supported by mean, standard deviatio, minimum, maximum, 25% percentile, 50% percentile, 75% percentile etc.

* We can see that the Minimum value is 0 and Maximum value is 433014, there is huge difference so both mean and median function would not work

* Also, there is a huge difference in the 25th, 50th, and 75th percentiles indicating that if we impute the values for Landsize column we will most probably introduce bias into the data.

* It is most appropriate to remove the rows where we find null values for Landsize.

In [13]:
# removing  all the rows where there is a null value in the landsize column
data = data[~np.isnan(data['Landsize'])]

# lets check the summary again
round(100*(data.isnull().sum()/len(data.index)), 2)

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         0.00
Bathroom         0.01
Car              0.46
Landsize         0.00
Lattitude        0.16
Longtitude       0.16
Regionname       0.00
Propertycount    0.00
dtype: float64

* We can see that now we have very low fraction of missing values in the columns Bathroom, Car, Lattitude, and Longitude
* Lets check the range values of Lattitude and Longitude.

In [55]:
# describing the values for lattitude and longitude
data[['Lattitude','Longtitude']].describe()

Lattitude    Longtitude
count  13581.000000  13581.000000
mean     -37.809204    144.995221
std        0.079257      0.103913
min      -38.182550    144.431810
25%      -37.856820    144.929600
50%      -37.802360    145.000100
75%      -37.756400    145.058320
max      -37.408530    145.526350

### Imputing using fillna() method

In [14]:
# as there is minute difference between the values of min, mean, max, 25th, 50th, and 75th percentile. 
# so we can use mean function to impute the values in lattitude and longtitude.

data['Lattitude'] = data['Lattitude'].fillna(data['Lattitude'].mean())
data['Longtitude'].fillna(data['Longtitude'].mean(), inplace = True)

In [58]:
# lets check the values present in the Bathroom Column

data['Bathroom'].value_counts()

1.0    7517
2.0    4987
3.0     921
4.0     106
0.0      34
5.0      28
6.0       5
8.0       2
7.0       2
Name: Bathroom, dtype: int64

In [59]:
# lets also check the values present in the Car Column

data['Car'].value_counts()

2.0     5606
1.0     5515
0.0     1026
3.0      748
4.0      507
5.0       63
6.0       54
8.0        9
7.0        8
10.0       3
9.0        1
Name: Car, dtype: int64

In [61]:
# we have to impute the values in Categorical Columns using the Mode function

data['Bathroom'].fillna(data['Bathroom'].mode()[0], inplace = True)
data['Car'].fillna(data['Car'].mode()[0], inplace = True)

In [62]:
# lets check the missing values present in the data

data.isnull().sum().sum()

0

we can see that there no null values left in the dataset i.e., we are done with this tutorial on Handling Missing Values present in Data.